File: Typescript/fraction.ts
class Fraction {
    readonly numerator: number
    readonly denominator: number

    constructor(numerator: number, denominator: number) {
        let d = this._gcd(numerator, denominator)
        this.numerator = numerator / d
        this.denominator = denominator / d
    }

    _gcd(a: number, b: number) : number {
        if (a === b) {
            return a
        } else {
            if (a > b) {
                return this._gcd(a-b, b)
            } else {
                return this._gcd(a, b-a)
            }
        }
    }

    public add(f : Fraction) : Fraction {
        return new Fraction(this.denominator*f.numerator+f.denominator*this.numerator, this.denominator*f.denominator)
    }

    public mult(f: Fraction) : Fraction {
        return new Fraction(this.numerator*f.numerator, this.denominator*f.denominator)
    }
    public toString() : string {
        return `${this.numerator} / ${this.denominator}`
    }
}

let half = new Fraction(1, 2)
let twothird = new Fraction(2, 3)
let onethird = new Fraction(1, 3)
let f = new Fraction(6, 76)

console.log(half.add(twothird).toString())
console.log(half.mult(twothird).toString())
console.log(half.mult(onethird).toString())
console.log(f.toString())

File: Scala/fraction.scala
class Fraction(var nominator: Int, var denominator: Int) {
  val gcd = _gcd(nominator, denominator)
  nominator = nominator/gcd
  denominator = denominator/gcd

  def this() = this(1, 1)

  def _gcd(a: Int, b: Int): Int =
    if (a == b)
      a
    else if (a > b)
      _gcd(a-b, b)
    else
      _gcd(a, b-a)

  def add(other: Fraction): Fraction =
    new Fraction(nominator*other.denominator + other.nominator*denominator, denominator*other.denominator)

  def mul(other: Fraction): Fraction =
    new Fraction(nominator*other.nominator, denominator*other.denominator)

  override def toString() = s"$nominator / $denominator"
}


object Compute extends App {
  val half = new Fraction(1, 2)
  val onethird = new Fraction(1, 3)
  val twothird = new Fraction(2, 3)
  println(half)

  val f1 = half.add(onethird)
  println(f1)

  val f2 = half.add(twothird)
  println(f2)

  val f3 = new Fraction(6, 76)
  println("6 / 76 = ", f3)

  val f4 = half.mul(twothird)
  println(f4)
}

